### PR TITLE
OCTOPUS-227 Support serviceCBranch,serviceC &  minorC as the variables

### PR DESCRIPTION
**Pull request by ozonophore**

### CODE REVIEW

File: src/main/kotlin/org/octopusden/releng/versions/KotlinVersionFormatter.kt [hunk 1]
```diff
@@ -3,16 +3,19 @@ package org.octopusden.releng.versions
 import org.octopusden.utils.expandContext
 import org.octopusden.utils.offsetFormat
 
-class KotlinVersionFormatter : VersionFormatter {
+class KotlinVersionFormatter(
+    private val serviceBranch : String = "serviceCBranch",
```
[the remark this answers is not in the document]
Comment (ozonophore) [reply src/main/kotlin/org/octopusden/releng/versions/KotlinVersionFormatter.kt]: done

File: src/main/java/org/octopusden/releng/versions/ComponentVersionFormatMatcher.java [hunk 2]
```diff
@@ -2,38 +2,46 @@
...
+    }
 
     public boolean matchesMajorVersionFormat(ComponentVersionFormat componentVersionFormat, String version) {
-        return FORMATTER.matchesFormat(componentVersionFormat.getMajorVersionFormat(), version);
+        return formetter.matchesFormat(componentVersionFormat.getMajorVersionFormat(), version);
     }
 
     public boolean matchesReleaseVersionFormat(ComponentVersionFormat componentVersionFormat, String version) {
-        return FORMATTER.matchesFormat(componentVersionFormat.getReleaseVersionFormat(), version);
+        return formetter.matchesFormat(componentVersionFormat.getReleaseVersionFormat(), version);
```
Comment (aryabokon) [inline src/main/java/org/octopusden/releng/versions/ComponentVersionFormatMatcher.java:20]: `formatter`

File: src/main/java/org/octopusden/releng/versions/NumericVersion.java [hunk 3]
```diff
@@ -14,15 +14,22 @@ public final class NumericVersion implements IVersionInfo {
     private final String rawVersion;
     private final boolean snapshot;
     private final boolean rcVersion;
+    private final VersionNames versionNames;
 
-    private NumericVersion(List<Integer> items, String rawVersion, boolean snapshot, boolean rcVersion) {
+    private NumericVersion(List<Integer> items,
```
Comment (aryabokon) [inline src/main/java/org/octopusden/releng/versions/NumericVersion.java:19]: ниже по коду, по понятным причинам, параметр `VersionNames versionNames` идет первым. Я бы его старался добавить везде первым

File: src/test/java/org/octopusden/releng/versions/NumericVersionTest.java [hunk 4]
```diff
@@ -10,10 +10,12 @@ class NumericVersionTest {
     private static final String STR_V_1_2_3_4 = "1.2.3.4";
     private static final String STR_V_1_2_3 = "1.2.3";
     private static final IVersionInfo VERSION_1_2_3_4 = version(STR_V_1_2_3_4);
-    private static final IVersionInfo VERSION_1_2_3 = NumericVersion.parse(STR_V_1_2_3);
+
+    private static final VersionNames VERSION_NAMES = new VersionNames("service", "serviceC", "minor");
```
Comment (aryabokon) [inline src/test/java/org/octopusden/releng/versions/NumericVersionTest.java:14]: тут нет ошибки в параметрах?

("serviceCBranch", "serviceC", "minorC")

File: src/main/java/org/octopusden/releng/versions/NumericVersion.java (hunk 3 above)
Comment (ozonophore) [reply src/main/java/org/octopusden/releng/versions/NumericVersion.java:19]: fixed

File: src/main/java/org/octopusden/releng/versions/ComponentVersionFormatMatcher.java (hunk 2 above)
Comment (ozonophore) [reply src/main/java/org/octopusden/releng/versions/ComponentVersionFormatMatcher.java]: fixed

File: src/main/java/org/octopusden/releng/versions/ComponentVersionFormatMatcher.java [hunk 5]
```diff
@@ -2,38 +2,46 @@
 
 public final class ComponentVersionFormatMatcher {
 
-    private static final KotlinVersionFormatter FORMATTER = new KotlinVersionFormatter();
+    private final VersionNames versionNames;
```
[the remark this answers is not in the document]
Comment (ozonophore) [reply src/main/java/org/octopusden/releng/versions/ComponentVersionFormatMatcher.java:5]: На 12 строке

File: src/main/java/org/octopusden/releng/versions/NumericVersion.java [hunk 6]
```diff
@@ -14,51 +14,18 @@ public final class NumericVersion implements IVersionInfo {
     private final String rawVersion;
     private final boolean snapshot;
     private final boolean rcVersion;
+    private final VersionNames versionNames;
 
-    private NumericVersion(List<Integer> items, String rawVersion, boolean snapshot, boolean rcVersion) {
+    public NumericVersion(VersionNames versionNames,
```
[the remark this answers is not in the document]
Comment (ozonophore) [reply src/main/java/org/octopusden/releng/versions/NumericVersion.java]: fixed

File: src/main/java/org/octopusden/releng/versions/ComponentVersionFormatMatcher.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (ozonophore) [reply src/main/java/org/octopusden/releng/versions/ComponentVersionFormatMatcher.java]: deleted

File: src/main/java/org/octopusden/releng/versions/NumericVersionFactory.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+    public IVersionInfo create(int... elements) {
+        StringBuilder stringBuilder = new StringBuilder();
+        for (int element : elements) {
+            if (stringBuilder.length() > 0) {
+                stringBuilder.append(".");
+            }
+            stringBuilder.append(element);
+        }
+        return create(stringBuilder.toString());
```
Comment (CamaradeRoman) [inline src/main/java/org/octopusden/releng/versions/NumericVersionFactory.java:51]: Joining raw version to string and splitting it back again is a worthless work.

File: src/main/java/org/octopusden/releng/versions/NumericVersionFactory.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    public NumericVersionFactory(VersionNames versionNames) {
+        this.versionNames = versionNames;
+    }
+
+    public IVersionInfo create(String rawVersion) {
+        Objects.requireNonNull(rawVersion, "version can't be null");
+        final List<String> strings = new ArrayList<>();
+        int l = 0;
+        int r;
+        for (r = 0; r < rawVersion.length(); r++) {
```
Comment (CamaradeRoman) [inline src/main/java/org/octopusden/releng/versions/NumericVersionFactory.java:20]: Is it better to use String.split with regexp?

File: src/main/java/org/octopusden/releng/versions/NumericVersionFactory.java [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+        strings.add(rawVersion.substring(l, r));
+
+        ArrayList<Integer> items = new ArrayList<>();
+        for (String stringItem : strings) {
+            int item;
+            try {
+                item = Integer.parseInt(stringItem);
+            } catch (NumberFormatException e) {
+                continue;
+//                item = 0;
```
Comment (CamaradeRoman) [inline src/main/java/org/octopusden/releng/versions/NumericVersionFactory.java:36]: ???

File: src/main/java/org/octopusden/releng/versions/VersionRange.java [hunk 10]
```diff
@@ -1,236 +1,12 @@
 package org.octopusden.releng.versions;
 
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.List;
-import java.util.Objects;
-
 /**
  * Simple implementation maven based version range specification.
```
Comment (CamaradeRoman) [inline src/main/java/org/octopusden/releng/versions/VersionRange.java:4]: This is not an implementation. Move the comment to VersionRangeImpl.

File: src/main/java/org/octopusden/releng/versions/NumericVersionFactory.java (hunk 9 above)
Comment (ozonophore) [reply src/main/java/org/octopusden/releng/versions/NumericVersionFactory.java:36]: i didn't touch it. Legacy code

File: src/main/java/org/octopusden/releng/versions/NumericVersionFactory.java (hunk 8 above)
Comment (ozonophore) [reply src/main/java/org/octopusden/releng/versions/NumericVersionFactory.java:20]: i didn't touch it. Legacy code

File: src/main/java/org/octopusden/releng/versions/VersionRange.java (hunk 10 above)
Comment (ozonophore) [reply src/main/java/org/octopusden/releng/versions/VersionRange.java:4]: This block exists in VersionRangeImpl

File: src/main/java/org/octopusden/releng/versions/NumericVersionFactory.java (hunk 9 above)
Comment (CamaradeRoman) [reply src/main/java/org/octopusden/releng/versions/NumericVersionFactory.java:36]: Does this commented string contains any valuable info?

**Review by CamaradeRoman — APPROVED** [review]

**Pull request merged**